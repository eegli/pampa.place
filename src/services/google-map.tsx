import {unsafeToggleHTMLElement} from '@/utils/misc';

export const GmapContainer = () => {
  return (
    <div id="__GMAP__CONTAINER__" data-testid="__GMAP__CONTAINER__">
      <div id="__GMAP__" data-testid="__GMAP__" style={{height: '100%'}} />
    </div>
  );
};

export class Gmap {
  static _map: google.maps.Map | undefined;

  static get div() {
    return document.getElementById('__GMAP__')!;
  }
  static get container() {
    return document.getElementById('__GMAP__CONTAINER__')!;
  }

  static get map() {
    if (!Gmap._map) {
      Gmap._map = new google.maps.Map(Gmap.div);
      console.log('Created new global Map instance');
    }
    return Gmap._map;
  }

  static toggle(tempContainer: Parameters<typeof unsafeToggleHTMLElement>[2]) {
    const undo = unsafeToggleHTMLElement(
      Gmap.div,
      Gmap.container,
      tempContainer
    );
    return () => {
      undo();
    };
  }
}
