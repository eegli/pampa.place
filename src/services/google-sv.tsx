import {unsafeToggleHTMLElement} from '@/utils/misc';

export const GstvContainer = () => {
  return (
    <div id="__GSTV__CONTAINER__" data-testid="__GSTV__CONTAINER__">
      <div id="__GSTV__" style={{height: '100%'}} />
    </div>
  );
};

export class Gstv {
  static get div() {
    return document.getElementById('__GSTV__')!;
  }
  static get container() {
    return document.getElementById('__GSTV__CONTAINER__')!;
  }
  private static _stv: google.maps.StreetViewPanorama | undefined;

  static get stv() {
    if (!Gstv._stv) {
      Gstv._stv = new google.maps.StreetViewPanorama(Gstv.div);
      console.log('%cCreated new global Street View instance', 'color: green');
    }
    return Gstv._stv;
  }

  static toggle<T extends HTMLElement>(tempContainer: T) {
    const undo = unsafeToggleHTMLElement(
      Gstv.div,
      Gstv.container,
      tempContainer
    );
    return () => {
      undo();
    };
  }
}
