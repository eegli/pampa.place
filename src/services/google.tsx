import {unsafeToggleHTMLElement} from '@/utils/misc';

export const GstvContainer = () => {
  return (
    <div id="__GSTV__CONTAINER__" data-testid="__GSTV__CONTAINER__">
      <div id="__GSTV__" data-testid="__GSTV__" style={{height: '100%'}} />
    </div>
  );
};

export const GmapContainer = () => {
  return (
    <div id="__GMAP__CONTAINER__" data-testid="__GMAP__CONTAINER__">
      <div id="__GMAP__" data-testid="__GMAP__" style={{height: '100%'}} />
    </div>
  );
};

export class StreetViewService {
  static get div() {
    return document.getElementById('__GSTV__')!;
  }
  static get container() {
    return document.getElementById('__GSTV__CONTAINER__')!;
  }
  private static _sv: google.maps.StreetViewPanorama | undefined;

  static get sv() {
    if (!StreetViewService._sv) {
      StreetViewService._sv = new google.maps.StreetViewPanorama(
        StreetViewService.div
      );
      console.log('%cCreated new global Street View instance', 'color: green');
    }
    return StreetViewService._sv;
  }

  static mount<T extends HTMLElement>(tempContainer: T) {
    const unmount = unsafeToggleHTMLElement(
      StreetViewService.div,
      StreetViewService.container,
      tempContainer
    );
    return () => {
      unmount();
    };
  }
}

export class MapService {
  static _map: google.maps.Map | undefined;
  static markers: google.maps.Marker[] = [];

  static addMarker(m: google.maps.Marker) {
    MapService.markers.push(m);
    return m;
  }
  // https://developers.google.com/maps/documentation/javascript/markers#remove
  static clearMarkers() {
    MapService.markers.forEach(m => m.setMap(null));
    MapService.markers.length = 0;
  }

  static get div() {
    return document.getElementById('__GMAP__')!;
  }
  static get container() {
    return document.getElementById('__GMAP__CONTAINER__')!;
  }

  static get map() {
    if (!MapService._map) {
      MapService._map = new google.maps.Map(MapService.div);
      console.log('%cCreated new global Map instance', 'color: green');
    }
    return MapService._map;
  }

  static mount<T extends HTMLElement>(tempContainer: T) {
    const unmount = unsafeToggleHTMLElement(
      MapService.div,
      MapService.container,
      tempContainer
    );

    return () => {
      unmount();
    };
  }
}
