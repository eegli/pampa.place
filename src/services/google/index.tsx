export {GmapContainer, GstvContainer} from './dom';
export {MapService, StreetViewService} from './static';
