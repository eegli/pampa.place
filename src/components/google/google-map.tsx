import {MapService} from '@/services/google';
import Box from '@mui/material/Box';
import {memo, ReactNode, useEffect, useRef} from 'react';

type BoundsProps = {
  bounds: {
    SW: google.maps.LatLngLiteral;
    NE: google.maps.LatLngLiteral;
  };
  padding?: number;
};

type CenterProps = {
  center: google.maps.LatLngLiteral;
  zoom?: number;
};

export type GoogleMapProps = {
  id?: string;
  children?: ReactNode;
  onMount?: (map: google.maps.Map) => unknown;
  onUnmount?: (map: google.maps.Map) => unknown;
} & (BoundsProps | CenterProps);

const _GoogleMap = ({
  children,
  id = 'google-map',
  onMount,
  onUnmount,
  ...rest
}: GoogleMapProps) => {
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      const unmountMap = MapService.mount(ref.current);
      console.info('%cGOOGLE MAP MOUNT', 'color: yellow');

      if ('bounds' in rest) {
        /* Order in constructor is important! SW, NE  */
        const bounds = new google.maps.LatLngBounds(
          rest.bounds.SW,
          rest.bounds.NE
        );
        google.maps.event.addListenerOnce(MapService.map, 'idle', () => {
          MapService.map.fitBounds(bounds, rest.padding || 0);
        });
      } else {
        MapService.map.setCenter(rest.center);
        MapService.map.setZoom(rest.zoom || 0);
      }

      if (typeof onMount === 'function') {
        onMount(MapService.map);
      }

      return () => {
        // Any cleanup logic
        if (typeof onUnmount === 'function') {
          onUnmount(MapService.map);
        }

        unmountMap();
        console.info('%cGOOGLE MAP UNMOUNT', 'color: yellow');
      };
    }
  }, [ref, onMount, onUnmount, rest]);

  return (
    <Box id={id} data-testid={id} ref={ref} height="100%" width="100%">
      {children}
    </Box>
  );
};

export const GoogleMap = memo(_GoogleMap, (prev, next) => prev.id === next.id);
