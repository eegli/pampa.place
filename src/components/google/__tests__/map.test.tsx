import {render, screen} from '@/tests/test-utils';
import {Map} from '@googlemaps/jest-mocks';
import React from 'react';
import {Gmap} from '../../../services/google-map';
import GoogleMap, {MapMode} from '../google.map';

jest.spyOn(React, 'useRef').mockReturnValue({
  current: document.createElement('div'),
});

describe('Google Map', () => {
  it('renders google map', () => {
    render(<GoogleMap activeMapId="1mSRVyWP3tLQ" />);

    expect(Gmap.map.fitBounds).toHaveBeenCalledTimes(1);

    expect(screen.getByTestId('__GMAP__CONTAINER__')).toBeInTheDocument();
    expect(screen.getByTestId('__GMAP__')).toHaveStyle('height:100%');
  });
  it('has preview mode', () => {
    const addGeoJsonSpy = jest.spyOn(Map.prototype.data, 'addGeoJson');
    const setStyleSpy = jest.spyOn(Map.prototype.data, 'setStyle');
    const removeGeoJsonSpy = jest.spyOn(Map.prototype.data, 'remove');
    const setOptonsSpy = jest.spyOn(Gmap.map, 'setOptions');

    const {unmount} = render(
      <GoogleMap activeMapId="1mSRVyWP3tLQ" mode={MapMode.PREVIEW} />
    );
    expect(setOptonsSpy.mock.calls[0]).toMatchSnapshot('preview settings');
    expect(addGeoJsonSpy).toHaveBeenCalledTimes(1);
    expect(setStyleSpy).toHaveBeenCalledTimes(1);
    unmount();
    expect(removeGeoJsonSpy).toHaveBeenCalledTimes(3);
  });

  it('has play mode', () => {
    const eventSpy = jest.spyOn(google.maps.event, 'removeListener');

    const {unmount} = render(
      <GoogleMap activeMapId="1mSRVyWP3tLQ" mode={MapMode.PLAY} />
    );

    unmount();
    expect(eventSpy).toHaveBeenCalledTimes(1);
  });
});
