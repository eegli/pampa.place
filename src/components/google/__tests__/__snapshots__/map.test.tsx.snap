// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Google Map has play mode: 1 1`] = `
<div>
  <div
    class="MuiContainer-root css-vg1u2y-MuiContainer-root"
    id="mother"
  >
    <div
      data-testid="__GMAP__CONTAINER__"
      id="__GMAP__CONTAINER__"
    />
    <div
      data-testid="__GSTV__CONTAINER__"
      id="__GSTV__CONTAINER__"
    >
      <div
        data-testid="__GSTV__"
        id="__GSTV__"
        style="height: 100%;"
      />
    </div>
    <div
      style="height: 100%;"
    >
      <div
        data-testid="__GMAP__"
        id="__GMAP__"
        style="height: 100%;"
      />
    </div>
  </div>
</div>
`;

exports[`Google Map has play mode: 2 1`] = `
<div>
  <div
    class="MuiContainer-root css-vg1u2y-MuiContainer-root"
    id="mother"
  >
    <div
      data-testid="__GMAP__CONTAINER__"
      id="__GMAP__CONTAINER__"
    />
    <div
      data-testid="__GSTV__CONTAINER__"
      id="__GSTV__CONTAINER__"
    >
      <div
        data-testid="__GSTV__"
        id="__GSTV__"
        style="height: 100%;"
      />
    </div>
    <div
      style="height: 100%;"
    >
      <div
        data-testid="__GMAP__"
        id="__GMAP__"
        style="height: 100%;"
      />
    </div>
  </div>
</div>
`;

exports[`Google Map has play mode: play settings 1`] = `
Array [
  Object {
    "clickableIcons": false,
    "disableDefaultUI": true,
    "draggableCursor": "crosshair",
    "gestureHandling": "auto",
    "mapTypeControl": true,
    "mapTypeControlOptions": Object {
      "mapTypeIds": Array [
        "hybrid",
        "roadmap",
      ],
    },
    "mapTypeId": "roadmap",
  },
]
`;

exports[`Google Map has play mode: update selected position 1`] = `
Object {
  "lat": 8,
  "lng": 8,
}
`;

exports[`Google Map has preview mode: preview settings 1`] = `
Array [
  Object {
    "clickableIcons": false,
    "disableDefaultUI": true,
    "draggableCursor": "crosshair",
    "gestureHandling": "none",
    "mapTypeControl": false,
    "mapTypeControlOptions": Object {
      "mapTypeIds": Array [
        "hybrid",
        "roadmap",
      ],
    },
    "mapTypeId": "roadmap",
  },
]
`;

exports[`Google Map has result mode: result settings 1`] = `
Array [
  Object {
    "clickableIcons": false,
    "disableDefaultUI": true,
    "draggableCursor": "crosshair",
    "gestureHandling": "auto",
    "mapTypeControl": true,
    "mapTypeControlOptions": Object {
      "mapTypeIds": Array [
        "hybrid",
        "roadmap",
      ],
    },
    "mapTypeId": "roadmap",
  },
]
`;
